File: json2excel.py
import pandas as pd
import json

def read_nested_json(json_file_path):
    try:
        with open(json_file_path, 'r') as file:
            data = json.load(file)
        return data
    except FileNotFoundError:
        print(f"Error: File not found - {json_file_path}")
        return None
    except json.JSONDecodeError:
        print(f"Error: Incorrect JSON format in file - {json_file_path}")
        return None

def flatten_json(json_data, parent_key='', separator='_'):
    flattened = {}
    if isinstance(json_data, dict):
        for key, value in json_data.items():
            new_key = f"{parent_key}{separator}{key}" if parent_key else key
            flattened.update(flatten_json(value, new_key, separator))
    elif isinstance(json_data, list):
        for i, item in enumerate(json_data):
            new_key = f"{parent_key}{separator}{i}" if parent_key else str(i)
            flattened.update(flatten_json(item, new_key, separator))
    else:
        flattened[parent_key] = json_data
    return flattened

def json_to_excel(json_file_path, excel_file_path):
    data = read_nested_json(json_file_path)

    if data:
        flattened_data = flatten_json(data)
        df = pd.DataFrame([flattened_data])
        df.to_excel(excel_file_path, index=False)
        print(f"Conversion successful! Excel file saved at {excel_file_path}")


# Example Usage:
json_file_path = '/home/nitin1053/Documents/python files/json py/input.json'
excel_file_path = '/home/nitin1053/Documents/python files/json py/input.xlsx'
json_to_excel(json_file_path, excel_file_path)

File: json2excel.js
const fs = require('fs');
const XLSX = require('xlsx');

function flattenObject(obj, parentKey = '') {
    let result = {};
    for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
            const newKey = parentKey ? `${parentKey}_${key}` : key;
            if (typeof obj[key] === 'object' && !Array.isArray(obj[key])) {
                result = { ...result, ...flattenObject(obj[key], newKey) };
            } else {
                result[newKey] = obj[key];
            }
        }
    }
    return result;
}

function addHyperlink(sheet, cell, sheetName) {
    const hyperlinkFormula = `=HYPERLINK("#'${sheetName}'!A1", "SHEET::${sheetName}")`;
    sheet[cell] = { f: hyperlinkFormula };
}

function jsonToExcel(jsonFile, excelFile) {
    try {
        const jsonData = JSON.parse(fs.readFileSync(jsonFile, 'utf8'));
        const workbook = XLSX.utils.book_new();

        for (const key in jsonData) {
            if (jsonData.hasOwnProperty(key)) {
                const sheetName = key.replace(/\s+/g, '_');
                const flattenedData = flattenObject(jsonData[key]);

                const ws = XLSX.utils.json_to_sheet([flattenedData]);
                XLSX.utils.book_append_sheet(workbook, ws, sheetName);

                // Add hyperlink in cell A1 to link to the corresponding sheet
                addHyperlink(ws, 'A1', sheetName);
            }
        }

        XLSX.writeFile(workbook, excelFile);

        console.log(`Conversion successful. Excel file saved as ${excelFile}`);
    } catch (error) {
        if (error.code === 'ENOENT') {
            console.error(`Error: File '${jsonFile}' not found.`);
        } else {
            console.error(`Error: ${error.message}`);
        }
    }
}

// Replace 'input.json' and 'output.xlsx' with your file names
jsonToExcel('input.json', 'output.xlsx');
